lib/services/embeddings.go

```diff
@@ -0,0 +1,396 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package services
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/trace"
+	jsoniter "github.com/json-iterator/go"
+	"gopkg.in/yaml.v3"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/ai"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+// Embeddings service is responsible for storing and retrieving embeddings in
+// the backend. The backend acts as an embedding cache. Embeddings can be
+// re-generated by an ai.Embedder.
+type Embeddings interface {
+	// GetEmbedding looks up a single embedding by its name in the backend.
+	GetEmbedding(ctx context.Context, kind, resourceID string) (*ai.Embedding, error)
+	// GetEmbeddings returns all embeddings for a given kind.
+	GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error)
+	// UpsertEmbedding creates or update a single ai.Embedding in the backend.
+	UpsertEmbedding(ctx context.Context, embedding *ai.Embedding) (*ai.Embedding, error)
+	// Embed takes a resource textual representation, checks if the resource
+	// already has an up-to-date embedding stored in the backend, and computes
+	// a new embedding otherwise. The newly computed embedding is stored in
+	// the backend.
+	Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error)
+}
+
+// PreciseJSONConfig is the fast JSON marshall/unmarshall configuration
+// we use in utils.FastMarshall() but with more precision on floats. This
+// is required to reduce the lost precision when storing embeddings in the
+// backend.
+var PreciseJSONConfig = jsoniter.Config{
+	EscapeHTML:                    false,
+	MarshalFloatWith6Digits:       false, // keep precision
+	ObjectFieldMustBeSimpleString: true,  // do not unescape object field
+	SortMapKeys:                   true,
+}.Froze()
+
```

Comment:
Sorry, I missed that earlier. Can we use the standard JSON library? We had many issues with this one in the past.